### PR TITLE
feat(tls): improve TryRecv logic and add test infra

### PR DESCRIPTION
**Pull request by glevkovich**
Production code:
- Implemented TlsSocket::TryRecv for robust TLS engine read/write/EOF
  handling, error propagation, and buffer management.
- Added TlsSocket::OnRecv to process RecvNotification variants, copy
  data to TLS engine input buffer, and invoke callbacks.
- Updated RegisterOnRecv and ResetOnRecvHook to delegate to next_sock_
  and support new OnRecv logic.
- Improved class structure for testability and buffer safety.

Testing:
- Added happy-flow tests ShortWrite and RegisterOnRecv in
  tls_socket_test.cc for TLS write/read and event-driven notification.
- Introduced tls_test_infra.h for reusable test/mocking infrastructure,
  enabling easier test composition and sharing.
- Added mock classes for TlsSocket and Engine, supporting strict and
  permissive behaviors for flexible unit testing.
- Implemented mock self-check tests for handshake and write ordering.
- Added parameterized test suite VerifyEdgeCases to simulate and
  validate edge cases for TlsSocket::TryRecv, ensuring robust error
  handling and concurrency safety.

### CODE REVIEW

File: util/tls/tls_socket.cc [hunk 1]
```diff
@@ -488,8 +521,135 @@ io::Result<size_t> TlsSocket::TrySend(const iovec* v, uint32_t len) {
...
+      if (recv_result) {
+        if ((*recv_result) > 0) {
+          engine_->CommitInput(*recv_result);
+          // Loop back to call engine_->Read() again and decrypt these bytes.
+          continue;
+        }
+        // *recv_result == 0 (Upstream socket EOF)
+        // The engine returned NEED_READ, but the socket is "dead".
+        // This is a "Dirty EOF" (TCP closed without TLS close_notify).
+        // We must return
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply util/tls/tls_socket.cc]: fixed

File: util/tls/tls_socket.cc [hunk 2]
```diff
@@ -488,8 +521,135 @@ io::Result<size_t> TlsSocket::TrySend(const iovec* v, uint32_t len) {
...
+    } else {
+      LOG(DFATAL) << "BUG: Unsupported read_result " << read_result;
+      return make_unexpected(make_error_code(errc::operation_not_permitted));
+    }
+  }
+  if (total_bytes_read > 0) {
+    DVSOCK(1) << "TryRecv returning " << total_bytes_read << " bytes";
+  }
+  return total_bytes_read;
 }
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply util/tls/tls_socket.cc]: false claim. single execution slot, fiber-atomic. We wont preemt.

File: util/tls/tls_socket.cc [hunk 3]
```diff
@@ -488,8 +521,135 @@ io::Result<size_t> TlsSocket::TrySend(const iovec* v, uint32_t len) {
...
+        // If we handled a NEED_WRITE successfully, we loop again to see what engine wants next
+        if (read_result == Engine::NEED_WRITE)
+          continue;
+      }
+      ///////////////////////////////////////////////////////////////
+      // 3. Check for read conflict:
+      // A read conflict implies the application is polling TryRecv while concurrently
+      // blocked on Recv, which is a usage error. We check this to prevent buffer corruption/crash.
+      if (read_in_progress) {
+        if (total_bytes_read > 0)
```
Comment (romange) [inline util/tls/tls_socket.cc]: you have this repeating pattern of 
```
if (total >0) 
  break;
 return ...
```

but you can just break, and then at line 652, check the status of total and either return total or the error.

File: util/tls/tls_socket.cc [hunk 4]
```diff
@@ -488,8 +521,135 @@ io::Result<size_t> TlsSocket::TrySend(const iovec* v, uint32_t len) {
...
+
+      // recv_result has an error (e.g. EAGAIN)
+      if (total_bytes_read > 0)
+        break;
+      return make_unexpected(recv_result.error());
+    } else if (read_result == Engine::EOF_STREAM) {  // case 3: Abrupt EOF
+      if (total_bytes_read > 0) {
+        break;
+      }
+      return make_unexpected(make_error_code(errc::connection_reset));
```
Comment (romange) [inline util/tls/tls_socket.cc]: why it is abrupt eof? 
we return recv return 0 in case the stream ended. similarly TryRecv should return 0 here.

File: util/tls/tls_socket.cc [hunk 5]
```diff
@@ -488,8 +521,135 @@ io::Result<size_t> TlsSocket::TrySend(const iovec* v, uint32_t len) {
...
+      // A read conflict implies the application is polling TryRecv while concurrently
+      // blocked on Recv, which is a usage error. We check this to prevent buffer corruption/crash.
+      if (read_in_progress) {
+        if (total_bytes_read > 0)
+          break;
+        return make_unexpected(make_error_code(errc::resource_unavailable_try_again));
+      }
+      ///////////////////////////////////////////////////////////////
+      // 4. Handle Pending Reads
+      auto input_buf{engine_->PeekInputBuf()};
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply util/tls/tls_socket.cc]: added :
`     DCHECK(PendingOutputSize() == 0) << "Pending output must be zero before reading input.";`

File: util/tls/tls_socket.cc [hunk 6]
```diff
@@ -488,8 +521,135 @@ io::Result<size_t> TlsSocket::TrySend(const iovec* v, uint32_t len) {
...
+          if (total_bytes_read > 0)
+            break;
+          return make_unexpected(send_result.error());
+        }
+        engine_->ConsumeOutputBuf(*send_result);
+
+        // If we couldn't send everything, we can't proceed safely.
+        if ((*send_result) < output_buf.size()) {
+          if (total_bytes_read > 0)
+            break;
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply util/tls/tls_socket.cc]: Added a warning.

File: util/tls/tls_socket.cc [hunk 7]
```diff
@@ -488,8 +521,135 @@ io::Result<size_t> TlsSocket::TrySend(const iovec* v, uint32_t len) {
...
+    // - NEED_WRITE: need to write to upstream socket
+    // - EOF_STREAM: connection closed abruptly by peer (no close_notify) / fatal TLS alert / system
+    // error / protocol violation
+    if (read_result > 0) {  // case 1:
+      buf.remove_prefix(read_result);
+      total_bytes_read += read_result;
+      continue;
+    } else if ((read_result == Engine::NEED_READ_AND_MAYBE_WRITE) ||
+               (read_result == Engine::NEED_WRITE)) {  // case 2: Handle NEED_READ/WRITE
+      ///////////////////////////////////////////////////////////////
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply util/tls/tls_socket.cc]: Not sure I understand why. And If I put it on master or my Pr branch it fail some unittrsts.

My logic:
- OpenSSL can detect a fatal error due to some corrupted records pushed to the engine.
- When that happen, On the SSL_read  it will set SSL_ERROR_SSL as an error  and return  result < 0
- But OpenSSL might still have other valid record(s) sitting in its internal buffers from a previous valid record that hasn't been drained yet by app.
- CHECK will fail

**Review by romange — APPROVED** [review]

**Review by romange — APPROVED** [review]
